util/src/main/java/org/datacommons/util/StringUtil.java

```diff
@@ -25,25 +25,34 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
 import java.util.regex.Pattern;
 import org.apache.commons.text.StringEscapeUtils;
 
 // Common set of utils to handle strings
 public class StringUtil {
   // From https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html
-  private static final List<String> DATE_PATTERNS =
-      List.of(
-          "yyyy",
-          "yyyy-MM",
-          "yyyyMM",
-          "yyyy-M",
-          "yyyy-MM-dd",
-          "yyyyMMdd",
-          "yyyy-M-d",
-          "yyyy-MM-dd'T'HH:mm",
-          "yyyy-MM-dd'T'HH:mm:ss",
-          "yyyy-MM-dd'T'HH:mm:ss.SSS",
-          "yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
+  // Pattern specification:
+  // https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html#appendPattern-java.lang.String-
+  // Key is the pattern, value is a list of lengths that this pattern could
+  // potentially match. This is used in filtering out impossible date checks by
+  // comparing the length of the input date string against the Set<Integer> here.
+  private static final Map<String, Set<Integer>> DATE_PATTERNS =
+      Map.ofEntries(
+          Map.entry("yyyy", Set.of(4)),
+          Map.entry("yyyy-MM", Set.of(7)),
+          Map.entry("yyyyMM", Set.of(6)),
+          Map.entry("yyyy-M", Set.of(6, 7)),
```

Comment:
Why is this 7? Won't that be handled by yyyy-MM ?

Ditto for line 50

Reply:
replying just for reference; the patterns do not map exactly 1:1 to characters they match; ie yyyy-M might match both 2022-7 and 2022-07.

Reply:
Yes, but lets make it match yyyy-MM.  That way we'll have 1 length per pattern, no?

Reply:
Is this resolved?

Reply:
I missed your earlier comment.

Currently, the two patterns that are breaking 1-length-per-pattern are `yyyy-M` and `yyyy-M-d`. These are both redundant if we are not interested in matching dates of the form `2022-7`/`2022-7-5` (no leading zero in single digit month and/or day components).

I have not deleted these redundant patterns since there are test cases for such cases, but I can certainly do so if supporting those dates is no longer a requirement.